networkx/algorithms/structuralholes.py

```diff
@@ -205,23 +249,52 @@ def constraint(G, nodes=None, weight=None):
     References
     ----------
     .. [1] Burt, Ronald S.
-           "Structural holes and good ideas".
-           American Journal of Sociology (110): 349–399.
+            "Structural holes and good ideas".
+            American Journal of Sociology (110): 349–399.
 
     """
+    import numpy as np
+    
     if nodes is None:
-        nodes = G
-    constraint = {}
-    for v in nodes:
-        # Constraint is not defined for isolated nodes
-        if len(G[v]) == 0:
-            constraint[v] = float("nan")
-            continue
-        constraint[v] = sum(
-            local_constraint(G, v, n, weight) for n in set(nx.all_neighbors(G, v))
+        # In order to compute constraint of all nodes,
+        # algorithms based on sparse matrices can be much faster
+        
+        # Obtain the adjacency matrix
+        P = nx.adjacency_matrix(G, weight=weight)
+        # Calculate mutual weights
+        mutual_weights = P + P.T
+        # Normalize mutual weights by row sums
+        sum_mutual_weights = mutual_weights.sum(axis=1)
+        val = np.asarray(
+            np.repeat(sum_mutual_weights, mutual_weights.getnnz(axis=0))
+        ).flatten()
+        mutual_weights.data = mutual_weights.data / val
+        # Calculate local constraints and constraints
+        local_constraints = (mutual_weights + mutual_weights @ mutual_weights).power(2)
+        constraints = ((mutual_weights>0).multiply(local_constraints)).sum(axis=1)
+        # Special treatment to isolated nodes
+        isolated_nodes = sum_mutual_weights == 0      # Mark isolated nodes
+        constraints[isolated_nodes] = float(
+            "nan"
+        )      # Constraint is undefined for isolated nodes
+        result = dict(
+            zip(list(G.nodes), 
+                np.asarray(constraints).flatten()
+               )
         )
-    return constraint
-
+        return result
+    
+    else:
```

Comment:
You can pull this code out of the `else` clause because the `if` clause has returned. Maybe put a comment like `# results for only requested nodes` to make it clear what the difference between `nodes is None` and not... :)

This looks quite good. Thanks. I haven't gone through it in detail yet, but it's quite promising.